### PR TITLE
Catchup

### PR DESCRIPTION
**Pull request by imerad**
This PR brings multiple changes to the project including : 
- added estimators Holland, Implicit (Lecué et al.) , GMOM
- added solvers GD, SGD, SVRG and SAGA
- Created mother class BaseLearner with BinaryClassifier and Regressor as subclasses
- added experiment script linreg_experiment.py for controlled linear regression
- added tracking mechanism to history

### CODE REVIEW

File: linlearn/_estimator.py [hunk 1]
```diff
@@ -79,6 +79,10 @@ def get_state(self):
     def partial_deriv_factory(self):
         pass
 
+    @abstractmethod
```
Comment (stephanegaiffas) [inline linlearn/_estimator.py:82]: En fait, on ne devrait pas mettre ces méthodes abstraites, mais elles devraient raise une erreur si on les appelle (pour les méthodes qui ne devraient pas être appelées, plutôt que de raise des erreurs dans les classes filles

File: linlearn/_estimator.py [hunk 2]
```diff
@@ -413,6 +453,333 @@ def partial_deriv(j, inner_products, state):
...
+
+        if self.fit_intercept:
+
+            @jit(**jit_kwargs)
+            def grad(inner_products, state):
+                deriv_samples = state.deriv_samples
+                deriv_samples_outer_prods = state.deriv_samples_outer_prods
+
+                gradient = np.zeros(X.shape[1] + 1)
+                for i in range(n_samples):
```
Comment (stephanegaiffas) [inline linlearn/_estimator.py:498]: Pourquoi on ne fait pas deux boucles imbriquées sans stocker deriv_samples ?
Aussi, si X.order == "C" on fait d'abord la boucle sur i pour j, alors que si X.order == "F" on double d'abord sur j et ensuite sur i

File: linlearn/_estimator.py [hunk 3]
```diff
@@ -413,6 +453,333 @@ def partial_deriv(j, inner_products, state):
...
+        X = self.X
+        y = self.y
+        loss = self.loss
+        value_loss = loss.value_factory()
+        deriv_loss = loss.deriv_factory()
+        n_samples_in_block = self.n_samples_in_block
+        n_blocks = self.n_blocks
+
+        # Better implementation of argmedian ??
+        @jit(**jit_kwargs)
```
Comment (stephanegaiffas) [inline linlearn/_estimator.py:567]: Ouille oui la on est en O(n) alors que ca devrait être O(log n)

File: linlearn/_loss.py [hunk 4]
```diff
@@ -50,27 +50,78 @@ def decision_function(w, out):
...
-            steps[j] = n_samples / (lip_const * col_j_squared_norm)
+            steps[j] = 1 / (max(median_of_means(X[:, j - 1] ** 2, block_size), 1e-8) * lip_const)
     else:
         steps = np.zeros(n_features, dtype=X.dtype)
         for j in prange(n_features):
-            col_j_squared_norm = 0.0
-            for i in range(n_samples):
-                col_j_squared_norm += X[i, j - 1] ** 2
-            steps[j] = n_samples / (lip_const * col_j_squared_norm)
+            steps[j] = 1 / (max(median_of_means(X[:, j] ** 2, block_size), 1e-8) * lip_const)
```
Comment (stephanegaiffas) [inline linlearn/_loss.py:124]: Mieux faut utiliser x * x que x ** 2 partout. Je ne suis pas sur que x ** 2 n'appelle pas pow(x, 2) qui est beaucoup plus lent

File: linlearn/_loss.py [hunk 5]
```diff
@@ -182,7 +233,7 @@ def __init__(self):
     def value_factory(self):
         @jit(**jit_kwargs)
         def value(y, z):
-            return 0.5 * (y - z) * (y - z)
```
Comment (stephanegaiffas) [inline linlearn/_loss.py:185]: Meme truc ici

File: linlearn/_loss.py [hunk 6]
```diff
@@ -50,27 +50,78 @@ def decision_function(w, out):
...
+            sum_block = 0.0
+
+    if last_block_size != 0:
+        block_means[n_blocks] = sum_block / last_block_size
+
+    mom = np.median(block_means)
+    return mom#, blocks_means
+
+@njit
+def steps_coordinate_descent(lip_const, X, block_size, fit_intercept):
```
Comment (stephanegaiffas) [inline linlearn/_loss.py:110]: Hmmm mais là on ne calcule que les pas via MOM... faudrait faire aussi une factory qui reçoit l'estimateur (ou le nom) et qui adapte les calculs en fonction ?

File: linlearn/_solver.py [hunk 7]
```diff
@@ -207,15 +316,500 @@ def cycle(coordinates, weights, inner_products, state_estimator):
...
+            def cycle(coordinates, weights, inner_products, state_estimator):
+                max_abs_delta = 0.0
+                max_abs_weight = 0.0
+                decision_function(X, fit_intercept, weights, inner_products)
+
+                grad = grad_estimator(
+                    inner_products, state_estimator
+                )
+                w_new = weights - step * grad
+                for j in range(1, n_weights):
```
Comment (stephanegaiffas) [inline linlearn/_solver.py:376]: Il me semble que j'avais mis une fonction qui penalize l'ensemble du vecteur

Comment (stephanegaiffas) [reply linlearn/_solver.py:376]: C'est dommage de ne pas mettre ces deux boucles ensemble (en démarrant à 1 et en gérant l'intercept à part)

**Review by stephanegaiffas — CHANGES_REQUESTED** [review]

File: linlearn/_solver.py [hunk 8]
```diff
@@ -207,15 +316,500 @@ def cycle(coordinates, weights, inner_products, state_estimator):
...
+        # apply_one_unscaled penalization function)
+        scaled_step = self.penalty.strength * self.step
+
+        if fit_intercept:
+
+            @jit(**jit_kwargs)
+            def cycle(coordinates, weights, inner_products, state_estimator):
+                max_abs_delta = 0.0
+                max_abs_weight = 0.0
+                decision_function(X, fit_intercept, weights, inner_products)
```
[the remark this answers is not in the document]
Comment (imerad) [reply linlearn/_solver.py:372]: Je me suis dit que ce serait lent d'appeler la fonction factory à chaque cycle. J'étais un peu embêté car la fonction factory est appelée dans solve() et la fonction retournée n'est pas directement accessible dans cycle. Je vais rajouter une condition spécifique à GD dans la boucle du solve dans la classe mère pour recalculer les produits scalaires après chaque cycle

File: linlearn/_estimator.py [hunk 9]
```diff
@@ -128,6 +132,37 @@ def partial_deriv(j, inner_products, state):
...
+        y = self.y
+        loss = self.loss
+        deriv_loss = loss.deriv_factory()
+        n_samples = self.n_samples
+
+        if self.fit_intercept:
+
+            @jit(**jit_kwargs)
+            def grad(inner_products, state):
+                gradient = np.zeros(X.shape[1] + 1)
```
[the remark this answers is not in the document]
Comment (imerad) [reply linlearn/_estimator.py:146]: In order to keep using a generic code for method solve() in the mother Solver class I'll allocate the gradient in the estimator state. If we do this through an "out" parameter we would need to override the solve() method, let me know if you prefer that anyway

File: linlearn/_estimator.py [hunk 10]
```diff
@@ -180,18 +213,20 @@ def partial_deriv(j, inner_products, state):
                 # Block counter
                 n_block = 0
                 if j == 0:
-                    for i in sample_indices:
```
[the remark this answers is not in the document]
Comment (imerad) [reply linlearn/_estimator.py:183]: done

File: linlearn/_estimator.py (hunk 2 above)
Comment (imerad) [reply linlearn/_estimator.py:498]: Si on ne stock pas les deriv_samples, il faudrait créer un array pour stocker tous les gradients par échantillon individuel ce qui correspond au volume du dataset (avec une colonne de plus s'il y a intercept) et je crois que c'est mieux de ne pas le faire. Pour maintenir cette approche efficace en mémoire l'ordre des boucles doit rester le même. Donc pour optimiser par rapport au type de contiguïté, je crois  qu'il faudrait plutôt redisposer les données avant de lancer le solver

File: linlearn/_estimator.py (hunk 3 above)
Comment (imerad) [reply linlearn/_estimator.py:567]: la fonction argpartition de numpy que j'utilisais avant n'est pas gérée par numba mais ce code est déjà très rapide, l'estimateur Implicit est plus rapide que tous les autres dans les expériences. La complexité est bien en O(n) mais on ne peut pas faire mieux (on a O(n) dès qu'on doit voir tous les éléments, O(log n) est seulement possible s'ils sont déjà triés)

File: linlearn/_loss.py (hunk 4 above)
Comment (imerad) [reply linlearn/_loss.py:124]: En fait x ** 2 n'appelle pas pow(x, 2) et à peu près aussi rapide au la multiplication x * x (j'ai vérifié avec timeit). J'ai utilisé x ** 2 surtout pour éviter de calculer des valeurs deux fois ou devoir les stocker

Comment (imerad) [reply linlearn/_loss.py:124]: J'ai fait des expériences avec timeit et c'est subtile, pour calculer le carré d'un vecteur (par coordonnées) c'est effectivement plus rapide de faire x*x mais si on calcule le carré d'une différence c'est plus rapide de faire (x - y)**2 plutôt que (x - y)*(x-y). J'adapte le code en conséquence

File: linlearn/_loss.py (hunk 5 above)
Comment (imerad) [reply linlearn/_loss.py:185]: corrigé

File: linlearn/_solver.py (hunk 7 above)
Comment (imerad) [reply linlearn/_solver.py:378]: J'ai fusionné les boucles

File: linlearn/_estimator.py [hunk 11]
```diff
@@ -128,6 +132,37 @@ def partial_deriv(j, inner_products, state):
...
+            def grad(inner_products, state):
+                gradient = np.zeros(X.shape[1])
+                for i in range(n_samples):
+                    gradient += deriv_loss(y[i], inner_products[i]) * X[i]
+                return gradient / n_samples
+
+            return grad
+
 
 ################################################################
```
[the remark this answers is not in the document]
Comment (imerad) [reply linlearn/_estimator.py:167]: Computing stochastic gradients is specific to SGD-type solvers. Adding such a method and plugging it into all the estimators poses new memory allocation issues and might be slower since it implies more function calls. Why not just manually compute sample gradients in the solvers that need this as we already do ?

File: linlearn/_estimator.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (imerad) [reply linlearn/_estimator.py:146]: adding the parameter "out" poses the question of where it should be allocated and since we use the hierarchy fit -> solve -> cycle and do not wish to allocate at each cycle, we would have to allocate in solve() but then we would have to pass this gradient place holder as a parameter to cycle which will be useless in many cases. To avoid this, I'll allocate it in the estimator state

[the remark this answers is not in the document]
Comment (stephanegaiffas) [reply linlearn/_estimator.py:146]: Yes, you should allocate it in the estimator state

File: linlearn/_estimator.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (stephanegaiffas) [reply linlearn/_estimator.py:167]: Yes but SGD, SVRG etc do need grad_sample somehow...

File: linlearn/_estimator.py [hunk 12]
```diff
@@ -301,6 +339,9 @@ def partial_deriv(j, inner_products, state):
 
             return partial_deriv
 
+    def grad_factory(self):
+        raise Exception("Don't use TMean estimator for full gradient")
```
[the remark this answers is not in the document]
Comment (imerad) [reply linlearn/_estimator.py:343]: done

File: linlearn/_estimator.py (hunk 2 above)
Comment (stephanegaiffas) [reply linlearn/_estimator.py:498]: Oui OK si on ne peut faire autrement, c'est OK

File: linlearn/_estimator.py (hunk 3 above)
Comment (stephanegaiffas) [reply linlearn/_estimator.py:567]: Oui tu as raison, autant pour moi

File: linlearn/_estimator.py (hunk 1 above)
Comment (imerad) [reply linlearn/_estimator.py:82]: J'enlève ces décorateurs pour partial_deriv_factory et grad_factory. On devrait probablement gérer les éventuelles mauvaises combinaisons solver-estimator directement à l'initialisation de BaseLearner

File: linlearn/_loss.py (hunk 4 above)
Comment (stephanegaiffas) [reply linlearn/_loss.py:124]: Je pense que c'est encore plus compliqué que ca. LLVM fait des optimisations et à mon avis x-y n'est calculé qu'une seule fois et X[i, j] n'est lu qu'un fois. Tu peux laisser comme ca, on pourra faire des comparaisons plus tard

**Review by stephanegaiffas — CHANGES_REQUESTED** [review]
Fait ces dernières corrections et ca sera bon.
Par contre, ce qu'il manque vraiment c'est des tests, a minima modifie les tests que j'avais fait pour tester que tous les solver trouvent des solutions identiques à celles de scikit-learn pour estimator='erm' et on ajoutera des tests pour les autres estimateurs plus tard

File: linlearn/_estimator.py [hunk 13]
```diff
@@ -238,6 +274,8 @@ def partial_deriv(j, inner_products, state):
 
             return partial_deriv
 
+    def grad_factory(self):
+        raise Exception("don't use mom with a full gradient solver (that's gmom)")
```
[the remark this answers is not in the document]
Comment (imerad) [reply linlearn/_estimator.py:278]: I added code that just fills a complete gradient by repeatedly calling partial_deriv from the same estimator. I will add warnings in the learner class for bad solver-estimator combinations

File: linlearn/_solver.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (imerad) [reply linlearn/_solver.py:372]: C'est fait

File: linlearn/_estimator.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (imerad) [reply linlearn/_estimator.py:167]: It is only a matter of one or two lines using deriv_loss() and accessing a row from the data X

File: linlearn/_loss.py (hunk 6 above)
Comment (imerad) [reply linlearn/_loss.py:110]: J'ai rajouté du code pour adapter le calcul des pas selon l'estimateur et le solveur

File: linlearn/_estimator.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (imerad) [reply linlearn/_estimator.py]: Added warnings in the BaseLearner class for strange estimator-solver combinations
